advisor/src/main/kotlin/advisors/VulnerableCode.kt

```diff
@@ -0,0 +1,168 @@
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.coroutineScope
+
+import org.ossreviewtoolkit.advisor.AbstractAdvisorFactory
+import org.ossreviewtoolkit.advisor.Advisor
+import org.ossreviewtoolkit.clients.vulnerablecode.VulnerableCodeService
+import org.ossreviewtoolkit.model.AdvisorDetails
+import org.ossreviewtoolkit.model.AdvisorResult
+import org.ossreviewtoolkit.model.AdvisorSummary
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.Vulnerability
+import org.ossreviewtoolkit.model.config.AdvisorConfiguration
+import org.ossreviewtoolkit.model.config.VulnerableCodeConfiguration
+import org.ossreviewtoolkit.model.createAndLogIssue
+import org.ossreviewtoolkit.model.utils.toPurl
+import org.ossreviewtoolkit.utils.collectMessagesAsString
+import org.ossreviewtoolkit.utils.showStackTrace
+
+/**
+ * An [Advisor] implementation that obtains security vulnerability information from a
+ * [VulnerableCode][https://github.com/nexB/vulnerablecode] instance.
+ */
+class VulnerableCode(name: String, config: AdvisorConfiguration) : Advisor(name, config) {
+    class Factory : AbstractAdvisorFactory<VulnerableCode>("VulnerableCode") {
+        override fun create(config: AdvisorConfiguration) = VulnerableCode(advisorName, config)
+    }
+
+    companion object {
+        /** The number of elements to request at once in a bulk request. */
+        private const val BULK_FETCH_SIZE = 100
+
+        /** A severity value used if no information is provided by VulnerableCode. */
+        private const val SEVERITY_UNDEFINED = -1f
+
+        /** A regular expression pattern to extract the ID of a vulnerability from its URI. */
+        private val vulnerabilityIdPattern = Regex(".*/api/vulnerabilities/([^/]+)/")
+
+        /**
+         * Try to extract the (internal VulnerableCode) ID from the given [url] of a vulnerability. Throw an
+         * exception if the URL has an unexpected form.
+         */
+        private fun extractVulnerabilityId(url: String): String =
+            vulnerabilityIdPattern.matchEntire(url)?.groups?.get(1)?.value
+                ?: throw IOException("Unexpected URL of a vulnerability: $url.")
+    }
+
+    private val vulnerableCodeConfiguration = config as VulnerableCodeConfiguration
+
+    override suspend fun retrievePackageVulnerabilities(packages: List<Package>): Map<Package, List<AdvisorResult>> {
+        val service = VulnerableCodeService.create(vulnerableCodeConfiguration.vulnerableCodeServerUrl)
+
+        val startTime = Instant.now()
+
+        val components = packages.map { it.purl }
+
+        @Suppress("TooGenericExceptionCaught")
+        return try {
+            val packagesToVulnerabilityRefs = mutableMapOf<String, VulnerableCodeService.Vulnerabilities>()
+
+            components.chunked(BULK_FETCH_SIZE).forEach { component ->
+                val requestResults =
+                    service.getPackageVulnerabilities(VulnerableCodeService.PackagesWrapper(component))
```

Comment:
This just fits in one line 😉

Reply:
Good spot. The line length is now exactly 120.